src/lang/interpreter/program-elements.ts

```diff
@@ -0,0 +1,62 @@
+import { ParserPosition } from "../parser";
+import { Interpreter } from "./interpreter";
+
+interface ProgramElementBase<Type extends string> {
+	readonly petype: Type;
+
+	/** The position where this program element starts in the source code */
+	where?: ParserPosition;
+
+	/**
+	 * Execute this program element
+	 *
+	 * @returns whether the program element is done. If false, it must be called again with an incremented instruction pointer.
+	 */
+	step(this: Interpreter): boolean;
+
+	/**
+	 * Determines whether the program element would be done after the next call to {@link step}.
+	 *
+	 * This controls when the debugger stops when stepping through a program.
+	 * Since the debugger wants to stop just before the PE's last instruction,
+	 * it halts when this method returns true.
+	 *
+	 * Some program elements don't want the debugger to halt on them and always return false
+	 * (e.g. by using `simfalse`).
+	 */
+	sim(this: Interpreter): boolean;
```

Comment:
I hope I got this right. I was initially confused that some PEs always return `false` here. So I think despite its name this method is actually only used to control when the debugger stops…

Reply:
Yes, you got that right. The expectation for `sim` is that it returns the same value as the next call to `step` will do, but without executing the step. As far as I remember, the only purpose is a nicer behavior of the debugger.